Mk ii hypo revisions(from poojawa)

crossexonar opened the pull request:
Adds in most of the revisions from: https://github.com/TS-Rogue-Star/Rogue-Star/pull/371

Adjustments Made:
Removed the round-start chem kits, order them from cargo instead.
Made all MkII Hypos compatible with the vials from the MKI hypo.
Gave more empty vials to kits to compensate (maybe too many)
Deluxe and Combat Hypospray Can load Bottles straight from the chem master, WOAH.

Spray mode now (currently) for RP flavor (some folks are afraid of needles yaknow?) ((topical medicine just isnt that common right now))
Spray/inject swap mode moved from alt click to a right click verb.
Alt click now lets you pick your injection rates -on- the hypospray without fiddling with the vials settings.

Mostly submitting for easy comparison between the two I guess.

crossexonar commented on code/datums/supplypacks/medical_vr.dm:
```diff
@@ -64,3 +64,18 @@
+/datum/supply_pack/med/mkiihypokits
+	name = "mk.II Hypospray Kits"
+	contains = list(
+				/obj/item/weapon/storage/firstaid/hypokit = 2, //RS Add
```
it was changed to contain the hypo + some empty vials and put into the vendors

Poojawa replied:
Considering I was the one who made this kit specifically an empty one, yes. I'm aware.

crossexonar commented on code/datums/supplypacks/medical_vr.dm:
```diff
@@ -64,3 +64,18 @@
 	access = access_medical_equip
 //Rs add end
+//RS Add MKII Hypokits
+/datum/supply_pack/med/mkiihypokits
+	name = "mk.II Hypospray Kits"
+	contains = list(
+				/obj/item/weapon/storage/firstaid/hypokit = 2, //RS Add
+				/obj/item/weapon/storage/firstaid/hypokit/brute = 2,
+				/obj/item/weapon/storage/firstaid/hypokit/burn = 2,
+				/obj/item/weapon/storage/firstaid/hypokit/antitoxin = 2,
+				/obj/item/weapon/storage/firstaid/hypokit/oxy = 2
+				)
+	cost = 25
```
Its cheap, cargo starts with 50 points, you can order 2 sets of everything to account for them being removed from the two nanomed vendors

crossexonar commented:
> So you're regressing the visual aspect of my hyposprays use for visual feedback for tiny vials that do not have any customization features. You get a single sprite rather than a player customizable one. Why?
> 
> ![2b8fc855dc2b9078226bf2238b054b39](https://private-user-images.githubusercontent.com/6293118/353838856-5024f99c-f2f4-4df5-b67c-42ac13c93713.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI0MzQwODksIm5iZiI6MTcyMjQzMzc4OSwicGF0aCI6Ii82MjkzMTE4LzM1MzgzODg1Ni01MDI0Zjk5Yy1mMmY0LTRkZjUtYjY3Yy00MmFjMTNjOTM3MTMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDczMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MzFUMTM0OTQ5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzdmOTY4MTFjNTYzYmMxZDdjMTI3ZTM1Nzg3MDczYTk0ZmU2ZTdkOTJkYzNiZmQzYTk3YjhlNDkzZWJlYzJmYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.7Ti6-UsdUndKuxxiEQcLjHNiMt9NFrNkv06TVmh89dM)
> 
> You're also locking these deliberately behind cargo-only ordering, which means they won't be ordered, won't be used, and are effectively removed from the code.
> 
> This is a terrible change and regression of the whole point of having it in the code.

Easy enough to revert, I commented it out not full on deleted it.

crossexonar commented on code/modules/reagents/reagent_containers/hypospraymkii.dm:
```diff
@@ -177,77 +185,101 @@
+	if(N)
+		amount_per_transfer_from_this = N
+
+/obj/item/weapon/hypospray_mkii/verb/swap_mode() //set amount_per_transfer_from_this
+	set name = "Swap spray mode"
+	set category = "Object"
+	set src in range(0)
+	switch(mode)
+		if(HYPO_SPRAY)
+			mode = HYPO_INJECT
+			to_chat(usr, "[src] is now set to inject contents on application.")
+		if(HYPO_INJECT)
+			mode = HYPO_SPRAY
+			to_chat(usr, "[src] is now set to spray contents on application.")
```
Alt click on most things that contain reagents sets the transfer rate, which is detailed just below this code.

Poojawa replied:
Except the applicator (The hypospray) doesn't contain reagents, This is intended to ensure there is no mixing of chemical payloads. Whatever it in the vial stays in the vial until used. That's the point. They are not syringes.

crossexonar commented on code/game/objects/items/weapons/storage/firstaid.dm:
```diff
@@ -168,68 +168,81 @@
-	desc = "Contains a Mk II Hypospray and its hypovials. It can only hold mk.II related items. This one is prefilled with Tricordazine."
-	starts_with = list(/obj/item/weapon/hypospray_mkii/tricord, /obj/item/weapon/reagent_containers/glass/bottle/hypovial/small/preloaded/tricordrazine = 3)
+	desc = "Contains a Mk II Hypospray and its hypovials. This one is prefilled with Tricordazine."
+	starts_with = list(/obj/item/weapon/hypospray_mkii/tricord, /obj/item/weapon/reagent_containers/glass/beaker/vial/preloaded/tricordrazine = 4)
```
Again, since they have to be ordered from cargo, they get one extra to round it out, Not a huge deal?

crossexonar commented on code/modules/reagents/reagent_containers/hypospraymkii.dm:
```diff
@@ -177,77 +185,101 @@
 	. = ..()
 	if(user.CanUseTopic(src, FALSE))
-		switch(mode)
+		if(possible_transfer_amounts && user.Adjacent(src))
+			set_APTFT()
```
The injection setting on the spray itself uses this

if(mode == HYPO_SPRAY)
		loadedvial.reagents.trans_to_mob(target, amount_per_transfer_from_this, CHEM_BLOOD)//CHEM_TOUCH) Changed to just be flavor
	else if(mode == HYPO_INJECT)
		loadedvial.reagents.trans_to_mob(target, amount_per_transfer_from_this, CHEM_BLOOD)

Poojawa replied:
Again. The hypospray is not a reagent_container. it doesn't actually hold any reagents.

crossexonar replied:
I select the injection amount, I inject monkey, monkey receives the selected amount of injection from the vial.
I just tested it and it works as intended?

crossexonar commented on code/game/objects/items/weapons/storage/firstaid.dm:
```diff
@@ -168,68 +168,81 @@
 	name = "Combat hypospray mk.II kit"
 	desc = "Contains a combat rated Deluxe Mk II Hypospray and its hypovials. It can only hold mk.II related items."
-	starts_with = list(/obj/item/weapon/hypospray_mkii/CMO/combat, /obj/item/weapon/reagent_containers/glass/bottle/hypovial/large/preloaded/combat = 4)
+	starts_with = list(/obj/item/weapon/hypospray_mkii/CMO/combat, /obj/item/weapon/reagent_containers/glass/beaker/vial = 10)
```
oh whoops that was ERT? Thought that one was CMO, reverting.

Poojawa replied:
it's a subtype of the CMO's because it's easier to have the same vars carried over. Says so in the name and description...

crossexonar commented on code/modules/reagents/reagent_containers/hypospraymkii.dm:
```diff
@@ -26,48 +26,53 @@
 /obj/item/weapon/hypospray_mkii/CMO
 	name = "hypospray mk.II deluxe"
-	allowed_containers = list(/obj/item/weapon/reagent_containers/glass/bottle/hypovial/small, /obj/item/weapon/reagent_containers/glass/bottle/hypovial/large)
+	allowed_containers = list(/obj/item/weapon/reagent_containers/glass/beaker/vial, /obj/item/weapon/reagent_containers/glass/bottle)
```
Its fat enough to fit variant 3 (whichever one chem masters default now to) of the bottles based on its shape

crossexonar commented on code/game/objects/items/weapons/storage/firstaid.dm:
```diff
@@ -168,68 +168,81 @@
+	///obj/item/weapon/reagent_containers/glass/beaker/vial/preloaded/bicaridine,
+	///obj/item/weapon/reagent_containers/glass/beaker/vial/preloaded/antitoxin,
+	///obj/item/weapon/reagent_containers/glass/beaker/vial/preloaded/dermaline,
+	///obj/item/weapon/reagent_containers/glass/beaker/vial/preloaded/dexalin //Go make chems silly
```
Roundstart chems outside of the base dylovene were expressed as non-desireable, Will ask if CMO version is to be included.

Poojawa replied:
But all the kits start prefilled for the same reason all the regular medikits do. 

the hypokits only hold hypo vials. 3 filled, one preload gives some very basic, quick use meds which should always be present in the stockroom anyway. I added them into the vendor to reduce map clutter.

crossexonar commented on code/modules/reagents/reagent_containers/hypospraymkii.dm:
```diff
@@ -78,7 +83,9 @@
 	icon_state = "combat2"
 	item_state = "combat2"
-	start_vial = /obj/item/weapon/reagent_containers/glass/bottle/hypovial/large/preloaded/combat
+	start_vial = /obj/item/weapon/reagent_containers/glass/beaker/vial
```
See above comment about the ERT one

crossexonar commented:
Also oh god this was meant to be a draft PR not a full PR, AAAAAAAAAA

BriarGarden commented:
Already said it elsewhere, saying it here too. I wholesale disagree with commenting out the new vials, and replacing them in all the kits. Need to _at the least_ add them to the main vendors, at this point including the large ones seeing as you've enabled adding whole bottles to some of them. Don't force your aesthetic on everyone else. Don't remove people's choices, just because you don't think they should have them.

Zoldorf commented:
How about adding the other vials in as just a sprite-state for the regular vial? Give it a context menu based proc that just allows you to select whichever vial you want, but code-wise they're all the same? Maybe add in a "large vial" sprite too to go along with the regular Vial for the large versions?

Poojawa commented:
> How about adding the other vials in as just a sprite-state for the regular vial? Give it a context menu based proc that just allows you to select whichever vial you want, but code-wise they're all the same? Maybe add in a "large vial" sprite too to go along with the regular Vial for the large versions?

Current hypovials already support this via editing with a pen. There would be nothing short of confusion by allowing completely different shapes and sizes in the visuals. the old vials are literally just test tubes meant for virology's usage. Meant for being used for blood products in the virology lab. Hypovials are visually distinct with colors and clear glass so people can immediately see what they're for - and the actual hypospray sprite itself uses them. 

Like the detective's gun has a lot of different visuals - but it's still a revolver. It's still revolver shaped and identifiable.

I can appreciate the idea of the compromise, but it's just a regression for the sake of removing content in a, frankly, heavy handed fashion.

crossexonar commented:
Reason this was meant to be a draft not a full PR was because I was just messing with the things to see if I -could- make it take the normal vials and bottles, but ended up having to replace some parts of the code because of the is_type() checks in it, and just started commenting things out untill the compiler stopped screaming at me.
The only -intended- changes here are:
moving the pre-filled kits to cargo for cheap.
Changing the spray toggle behaviour to a verb instead of a dedicated alt click (its rare enough to be used that it doesnt need a hotkey)
Making the spray mode be a flavor change as again, topical meds arnt -really- a thing here, if not taken, spray mode should have some better instructions so people dont use it incorrectly.

Adjusting injection amount ON the hypospray with alt click, yes it does actually work.

Compatability with the other vials and bottles from chem master for the relative Hyposprays to make them simpler/more enticing to use, as currently all the CMO's don't seem to care that the MK2's even exist, and just use syringe + bottles, Letting you load bottles straight into the CMO version would make it more likely to be actually used.
(chem removal from the CMO package is intended under the same logic of "move the pre-filled kits to cargo)
Otherwise the Mk1 Kit should have a prefilled set too, and use one of those openable case settings like the Captains sidearm choice box to pick which one you prefer.

Poojawa commented:
People are free to use whichever option suits their playstyle. If they are more comfortable with a particular set of tools, then it's silly to outright remove 'the competition' because it doesn't suit someone's particular playstyle. It doesn't need to be 'more enticing' to use. You're just removing options from people.

Removing the kits from the vendor is just soft-removing them from the game entirely. Locking them behind cargo, or science, or any other barrier means they will never be used. And I reject that outcome. There is no requirement for someone to play a particular way. There never should be that either. 

Changing the injection amount on the hypospray doesn't work because now there's no point to bottles at 60u because the hard cap is now 15u maximum. May as well just keep using pill force feeding or harm stabbing syringes or the medigun,  because you've missed the point about it just being the applicator, not the reagent container. You've also now completely ruined the ability to have fine-tuned loadouts, because now vials cannot be pre-set to a specific dosage and easily used, you now have to spend time adjusting the applicator every time you swap vials. Pop ups are bad for gameplay design - no one likes pop ups when they need to be playing the game.

Loading bottles was discouraged because it promoted lazier gameplay in general. My design choices were made to make something different, and not just yet another bottle carry device that simply removes the need to carry a syringe. The scarcity of the empty kits, hyposprays, and hypovials  was resolved in my tweaks, enabling the autolathe to produce more hypovials but not sprays meant that you couldn't just instantly make 15,000u of chem bottles at the chem master and never worry again. Haven't you mentioned how silly that was before? Why are you enabling the very behavior you've been grumpy about?

As for topical meds, I'm becoming motivated by pure spite to not only enable them but to ensure people have even more options to play with if they so choose. 

All in all, this PR simply removes player choice for seemingly little reasoning and I don't find that an acceptable choice.

FartMaster69420 commented:
How is moving something to cargo removing it from the game?

I see medical ordering clotting crates almost every shift.

Zoldorf commented:
That line of reasoning is entirely illogical, "Comfortable with a particular set of tools" the vast majority of people will use whatever is best for the job, and to scoff it up to mere preference is absurd.

And simply removing somethings round-start availability is NOT "soft-removing" them, it's doing the balancing that anything powerful should have, and that is to limit its availability, even if only a slight bit.

Also, the original Hypospray was quite literally made as a replacement for syringes, but was limited in availability, that is what the sole purpose of it was, deluxe syringe for the CMO, much like the deluxe tools for the CE.

You then take this and consider, how does it make any sense that this tool is now given out whilly nilly to every doctor but has been turned into a worse syringe that is more annoying to use?

-

You seem whole heartily obsessed with pushing what YOU want, you don't listen to feedback on anything, you ignore and downplay others opinions on everything, you constantly push and push and push your PR's everywhere you go, you constantly argue and bicker about anything and everything, like here in this very PR, instead of questioning intentions and trying to figure out why they actually want this and maybe coming to a solution that would be better, you instantly begin with snarky comments and constant sass with little reasoning for why something even needs to change, you go to just referring to things that YOU prefer as 'improvements' and anything others do as 'regressions' because you don't like it, what it does and means for the overall feel and gameplay is just not something you seem at all bothered with.

And now I see you start to say that others do "my way or no way" and "heavy handed ways" as you are the very person who went about making the Hypospray MK.2 which people didn't even ask for and nobody wants to use it, I haven't seen one used in forever because it's just a worse syringe, and now you come complaining that others who actually want to have a use for it are trying to fix it and make it less inconvenient to use without offering any sort of fixing or solutions.

I see this sort of behaviour more and more nowadays too, this constant desire to "code flex" and do all these massive big changes to things with zero regard for those who made things prior, you always word it as "improvements" and "fixes" when just scooping out whole chunks of code by others and replacing it with your poorly made and practically untested slop, I'm not new to the SS13 block, I've been around for over a decade by now and I've seen your kind do this over and over, worse yet, you don't even play the server that you push this onto either, you just want to feel as if you are important and cool, but you're just another code-ape who mashes their hands on their keyboard and try to brute force it into a merge the instant it compiles without errors, "Works on my machine" they say as literally nothing works, "oh no, this change is totally needed and MUST be added" they say as the playerbase screams in anguish, and the moment people give you actual push-back and question you, suddenly you pack your bags and leave, giving people some sob-story about being "unwanted" and trying to pull the heart strings and end up causing drama over it.

FartMaster69420 commented:
>As for topical meds, I'm becoming motivated by pure spite to not only enable them but to ensure people have even more options to play with if they so choose.

This really isn't a good mindset to build your PRs with.

Poojawa commented:
> Zoldorf: *text wall tirade*

Uh, sure buddy. I hope you find a better source of happiness than just screaming at me though. I can't fathom what personal wrong I did to you to have earned such hatred. Maybe consider taking a break?

Zoldorf commented:
That right there is my bloody problem, your endless dismissal, snarky comments, sass and enormous hypocrisy.

Poojawa commented:
> That right there is my bloody problem, your endless dismissal, snarky comments, sass and enormous hypocrisy.

If you say so!

Have the day you deserve.

Poojawa commented:
Not too sure about enabling quick swap on all of them. Just being on the CMO's would be sensible though.

My reasoning for that is staff head equipment should always be fundmentally better as a gameplay rule. HoS armor is better than regular sec, Cheng's gear is better, etc etc. Non deluxe versions should retain their requirement to slow swap

I'm not sure why you're using "intended" as a justification for the size alterations. 30u small hypovials are smaller than 60u beakers. That's what that means. in order to be equivalent to just a box of bottles with a syringe; it would mean at least 14 hypovials (with one preloaded in a hypospray) are needed.

crossexonar commented:
Intended as in, you in the comments said that it was meant to be smaller than a beaker, so i set its size to Tiny, not its volume.

crossexonar commented:
The deluxe version of the Hypospray already carries Double the capacity of the normal hypo's and can inject more at a time, and works faster, if you want these to actually compete with syringes hotswap is one of the ways that makes them a bit easier and less frustrating (I'd of set it so you could swap them ON the belt but that requires a frustrating amount of coding I don't want to mess with)

Poojawa commented:
They don't need to hot swap to be competitive. the deceased usage time is perfectly fine. As is the ability to carry the same amount of chems in a box at one point. 

hot swapping for all of them makes them *superior* because it lessens the amount of inventory juggle required. it also removes the point of having the Ctrl-click function. I reject making them too powerful because they shouldn't wholesale replace any other method. 

Remember, syringes have inventory juggle, as well as paying attention to being on draw/extract modes. And you have to fully empty what ever is unused. The advantages of it are being able to mix chemicals on the fly.

pills are high dosage, but require you to take off your mask and you can't change their premix. 

Hypovials fill a middle ground between both. You can more easily swap and tailor your medication, but you still have to be mindful of both hands and fiddle with enough inventory so it isn't medigun levels of brain dead easy. restricting hotswap to ERT or CMO provides that power but there's only the one, so you can't just outfit the entire medical department with 'em.

crossexonar commented:
Fair enough, Counterpoint is, syringe has no inventory management at all outside of tracking your inject/draw mode, you open your belt, syringe in your ear to your hand, touch bottle, touch patient, its why everyone uses them, theyre braindead easy and blow hyposprays of all kinds out of the water, I gave hotswap to the smaller models because it puts them at least on -par- with the syringes, without completely outshining them, similar inject times for 15u sprays to syringe (not hypothetical, ran testing with them all), I took away bottle support for the deluxe hypo because it was able to transfer 60u at once which was silly.

ShianShui commented:
Sari asked me to come playtest this, this does make the mk2 hypo more appealing. As it is, it's not something that the majority of medical players, including myself, find particularly useful. I would probably still prefer my syringe and mk1 hypo(as cmo) but I can see this making it a viable option to replace the syringes versatility.

keane0tts commented:
I have given the current mkII's a serious go, and always found them cumbersome and slow to use compared to a syringe. Frankly, the hyposprays in Star Trek, that these are, obviously, based off of, always seemed so cool to use. The ones in game? Not at all. 

I gave these a try in a PTR and at the very least, I liked how easily swapable the vials are. That change, alone, ends a lot of the gripes I have about the mkII's. I cannot stress enough how much I like that as a change. If nothing else from this got added but hot-swapability, I would be happy. I like the FANTASY of using hypos, but in reality, I just stick to syringes because they're frankly better in every regard to the mkII. 

I also quite liked the medical vendor having just one hypospray kit in it, and the person using it has to fill the vials with chems. Currently, I've seen people take one of each of the four different kits out and mix vials around, on multiple occasions creating a huge mess by not putting the unused kits back. Or they'll take one kit, throw out some of the vials, and then spam the vendor for empty vials. 

I would support just having a mkII kit that's just an empty hypo, and the rest of the slots filled with empty vials. That's what we're going to end up doing with it anyway, why make us spam a vendor for more empty vials? You can keep the number of kits per vendor at 4, too. We have enough vendors it doesn't matter.

Poojawa commented:
friendly reminder that my original PR was reopened prior to this #371 

It increases the usability of the hyposprays as well as increases empty kits in the vendor, while ensuring that cargo isn't the only source of the devices to begin with.

crossexonar commented:
Yes, I see that its open, this one is still open in the hopes that you take feedback from the community about what makes them more usable rather than just giving more kits.

crossexonar commented:
Fixed the issue of being able to stack injections one on top of another to all complete at the same time.
Adjusted the delay on injection to scale with how much you're injecting at a time.
Made it more obvious that you're selecting the vials transfer amount in the transfer list (formerly was just "0" Now is labled "vial")
Commented out some variables that were un-used

crossexonar commented:
Work and time put into this far exceeds the original scope of what I had intended, will open a new one later.

Voidastra commented:
### My take on the changes so far.

I tested some stuff on the PTR. I think that it is a good change, but doesn't and probably won't change my stance on using a syringe. If you want a good design I think that this is a step in the right direction but to reiterate what some people have said in this thread, it ultimately doesn't really give me a reason to swap. 

### My personal thoughts on the current Hypospray

- I'd rather have nice stuff gatekept behind cargo. science, etc. and have it be just a bit better than something else. That how current surgery tools, medical supplies, and chemicals work, and it seemingly creates a sense of progression. 
- Hyposprays, in their current state, should not be in vendors. There shouldn't be separate kits, just a blank one. I never see it get used except by untrained medical staff. This just causes clutter in medical, where we are already strapped for counter-top space.
- Despite the previous bullet point, I have see it getting used at start of round and someone gets hurt. It is nice to have a quick-fix to some things, but it only takes 10 seconds to get 120 units of whatever you need most of the time (like Bicar).


### Potential Other Solutions

- Allow for the hypospray to quick swap canisters from its bag by holding the hypospray and clicking. This will allow players to hold the bag in one hand and the hypospray in the other. Effectively this will allow medical to have quick access to important meds they prepare. 
- Put the hypospray behind science or cargo, and allow it to hold multiple cartridges that you can swap between like the MediGun. This will allow players to have it loaded and in the players pocket for fixes. 



### General Conclusion
The hypospray 2 proposed changes are great, but ultimately won't fix the issues at hand being: I'd rather stick to a syringe because it is more adaptable and easy to use. I can fix up a patient in a few clicks rather than going into my bag to grab another vial and loading the vial then setting the vial to the right dosage then applying. At it's current state, it needs more work if this wants to be QOL update. Hyposprays are a cool idea in theory but prove to be less than useful in many cases, not just the v2's. The standard one (v1 I believe) and the advanced (the one the CMO gets) injects instantly. There is/are another version(s) of the hypospray like the once science can make, but I am not entirely versed with those to make a comment. But regarding the v1 and the advanced, I see it used to inject some basics here and there, but overall again, it just doesn't do what you want. Personally the best use case I have found for it was loading it full of sleepy juice in those important moments. Although, while reusable, you can simply use an auto injector for the same effect. If you want the hypospray v2 to really be something that people might use, it needs to have a certain edge over a syringe, whether it be speed of injection, the ability to hot-swap from the hypo bag, the ability to hold more meds via vials, etc. The changes on the PTR are unfinished but if cleaned up, something that I'd like to see regardless of how this continues to get worked on.
